javascript/encoders_test.js

```diff
@@ -0,0 +1,233 @@
+    assert(existsSync(tokenizerPath), `Tokenizer file should exist at ${tokenizerPath}`);
+    for (const modalityPath of Object.values(modalityPaths)) {
+        assert(existsSync(modalityPath), `Modality file should exist at ${modalityPath}`);
+    }
+}
+
+async function testGetCheckpoint() {
+    console.log("- `testGetCheckpoint`: Start");
+
+    try {
+        const modalities = [Modality.TextEncoder, Modality.ImageEncoder];
+
+        for (const modelId of [
+            'unum-cloud/uform3-image-text-english-small',
+            'unum-cloud/uform3-image-text-english-base',
+            'unum-cloud/uform3-image-text-english-large',
+            'unum-cloud/uform3-image-text-multilingual-base',
+        ]) {
+            await tryGettingCheckpoint(modelId, modalities, hf_token);
+        }
+
+        console.log("- `testGetCheckpoint`: Success");
+    } catch (error) {
+        console.error("- `testGetCheckpoint`: Failed", error);
+    }
+}
+
+async function tryTextEncoderForwardPass(modelId) {
+    const modalities = [Modality.TextEncoder];
+    const { configPath, modalityPaths, tokenizerPath } = await getModel(
+        modelId,
+        modalities,
+        hf_token,
+        '.onnx'
+    );
+
+    const textProcessor = new TextProcessor(configPath, tokenizerPath);
+    await textProcessor.init();
+    const processedTexts = await textProcessor.process("a small red panda in a zoo");
+
+    const textEncoder = new TextEncoder(modalityPaths.text_encoder, textProcessor);
+    await textEncoder.init();
+    const textOutput = await textEncoder.encode(processedTexts);
+    assert(textOutput.embeddings.dims.length === 2, "Output should be 2D");
+
+    await textEncoder.dispose();
+}
+
+async function tryImageEncoderForwardPass(modelId) {
+    const modalities = [Modality.ImageEncoder];
+    const { configPath, modalityPaths } = await getModel(
+        modelId,
+        modalities,
+        hf_token,
+        '.onnx'
+    );
+
+    const imageProcessor = new ImageProcessor(configPath);
+    await imageProcessor.init();
+    const processedImages = await imageProcessor.process("assets/unum.png");
+
+    const imageEncoder = new ImageEncoder(modalityPaths.image_encoder, imageProcessor);
+    await imageEncoder.init();
+    const imageOutput = await imageEncoder.encode(processedImages);
+    assert(imageOutput.embeddings.dims.length === 2, "Output should be 2D");
+
+    await imageEncoder.dispose();
+}
+
+function cosineSimilarity(vecA, vecB) {
+    // We may be receiving a complex tesnor type, so let's check if it
+    // has an array member named `data`.
+    if (vecA.data) {
+        vecA = vecA.data;
+    }
+    if (vecB.data) {
+        vecB = vecB.data;
+    }
+
+    let dotProduct = 0.0;
+    let normA = 0.0;
+    let normB = 0.0;
+    for (let i = 0; i < vecA.length; i++) {
+        dotProduct += vecA[i] * 1.0 * vecB[i];
+        normA += vecA[i] * 1.0 * vecA[i];
+        normB += vecB[i] * 1.0 * vecB[i];
```

Comment:
Never seen this `* 1.0 *` trick before 👀 is it needed for type-casting or something?

Reply:
Yep, in compiled languages we sometimes cast code like that, but in JS it may be inefficient.